thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/classification/ClassificationTaskRunner.java

```diff
@@ -1,17 +1,161 @@
+
+  private long windowStart;
+  private long windowEnd;
+  private ClassificationConfigDTO classificationConfig;
+  private AnomalyFunctionFactory anomalyFunctionFactory;
+  private AlertFilterFactory alertFilterFactory;
+
+  Map<Long, AnomalyFunctionDTO> anomalyFunctionConfigMap = new HashMap<>();
+  Map<Long, AlertFilter> alertFilterMap = new HashMap<>();
+
   @Override
   public List<TaskResult> execute(TaskInfo taskInfo, TaskContext taskContext) throws Exception {
     List<TaskResult> taskResults = new ArrayList<>();
 
+    LOG.info("Setting up task {}", taskInfo);
+    setupTask(taskInfo, taskContext);
+
+    runTask(windowStart, windowEnd);
+
     return taskResults;
   }
+
+  private void setupTask(TaskInfo taskInfo, TaskContext taskContext) {
+    ClassificationTaskInfo classificationTaskInfo = (ClassificationTaskInfo) taskInfo;
+    windowStart = classificationTaskInfo.getWindowStartTime();
+    windowEnd = classificationTaskInfo.getWindowEndTime();
+    classificationConfig = classificationTaskInfo.getClassificationConfigDTO();
+    anomalyFunctionFactory = taskContext.getAnomalyFunctionFactory();
+    alertFilterFactory = taskContext.getAlertFilterFactory();
+  }
+
+  private void runTask(long windowStart, long windowEnd) {
+    long mainFunctionId = classificationConfig.getMainFunctionId();
+    addAnomalyFunctionAndAlertConfig(mainFunctionId);
+    AlertFilter alertFilter = alertFilterMap.get(mainFunctionId);
+
+    // Get the anomalies from the main anomaly function
+    List<MergedAnomalyResultDTO> mainAnomalies =
+        mergedAnomalyDAO.findAllOverlapByFunctionId(mainFunctionId, windowStart, windowEnd, false);
+    List<MergedAnomalyResultDTO> filteredMainAnomalies = filterAnomalies(alertFilter, mainAnomalies);
+
+    // Sort merged anomalies by the natural order of the end time
+    Collections.sort(filteredMainAnomalies, new MergeAnomalyEndTimeComparator());
+    // Run classifier for each dimension of the anomalies
+    ListMultimap<DimensionMap, MergedAnomalyResultDTO> updatedMainAnomaliesByDimension =
+        dimensionalShuffleAndUnifyClassification(filteredMainAnomalies);
+
+  }
+
+  private ListMultimap<DimensionMap, MergedAnomalyResultDTO> dimensionalShuffleAndUnifyClassification(
+      List<MergedAnomalyResultDTO> mainAnomalies) {
+
+    ListMultimap<DimensionMap, MergedAnomalyResultDTO> updatedMainAnomaliesByDimension = ArrayListMultimap.create();
+    // Terminate if the main anomaly function has not detected any anomalies in the given window
+    if (CollectionUtils.isEmpty(mainAnomalies)) {
+      return updatedMainAnomaliesByDimension;
+    }
+
+    // Sort anomalies by their dimensions
+    ListMultimap<DimensionMap, MergedAnomalyResultDTO> mainAnomaliesByDimension = ArrayListMultimap.create();
+    for (MergedAnomalyResultDTO mainAnomaly : mainAnomalies) {
+      mainAnomaliesByDimension.put(mainAnomaly.getDimensions(), mainAnomaly);
```

Comment:
It is initialized with an empty dimension map, but it could be null if someone accidentally sets it to null. That is still ok because mainAnomaliesByDimension takes null key; in this case, null key becomes a dimension. The classification works as long as anomalies from different dimensions do not be mixed together.